File: core/src/main/java/edu/knowitall/taggers/constraint/Constraint.java
package edu.knowitall.taggers.constraint;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.List;

import edu.knowitall.taggers.tag.ParseTagException;
import edu.knowitall.taggers.Type;
import edu.knowitall.tool.chunk.ChunkedToken;
import edu.knowitall.tool.stem.Lemmatized;

public abstract class Constraint {
    public abstract boolean apply(List<Lemmatized<ChunkedToken>> sentence, Type tag);

    public static Class<?> getConstraintClass(String classname) throws ClassNotFoundException {
        return (Class<?>)Class.forName(Constraint.class.getPackage().getName() + "." + classname);
    }

    public static Constraint create(String classname, Class<?>[] argTypes, Object[] argValues) throws ParseTagException {
        try {
            return create(getConstraintClass(classname), argTypes, argValues);
        } catch (ClassNotFoundException e) {
            throw new ParseTagException(e);
        }
    }

    public static Constraint create(Class<?> tagger, Class<?>[] argTypes, Object[] argValues) throws ParseTagException {
            try {
                Constructor<?> constructor = tagger.getConstructor(argTypes);
                return (Constraint)constructor.newInstance(argValues);
            } catch (SecurityException e) {
                throw new ParseTagException("Could not create class: " + tagger.getName(), e);
            } catch (NoSuchMethodException e) {
                throw new ParseTagException("Could not create class: " + tagger.getName(), e);
            } catch (InstantiationException e) {
                throw new ParseTagException("Could not create class: " + tagger.getName(), e);
            } catch (IllegalAccessException e) {
                throw new ParseTagException("Could not create class: " + tagger.getName(), e);
            } catch (InvocationTargetException e) {
                throw new ParseTagException("Could not create class: " + tagger.getName(), e);
            }
    }

    public static Constraint create(String classname) throws ParseTagException {
        return create(classname, new Class<?>[] {}, new Object[] {});
    }
}

File: core/src/main/scala/edu/knowitall/taggers/constraint/Constraint.scala
package edu.knowitall.taggers.constraint;

import java.lang.reflect.Constructor
import java.lang.reflect.InvocationTargetException
import java.util.List
import edu.knowitall.tool.typer.Type
import edu.knowitall.tool.chunk.ChunkedToken
import edu.knowitall.tool.stem.Lemmatized;
import edu.knowitall.repr.sentence.Sentence

abstract class Constraint[-S <: Sentence] {
  type TheSentence = S
  def apply(sentence: S, tag: Type): Boolean
}

object Constraint {
  def create[S <: Sentence](classname: String): Constraint[S] = {
    create[S](classname, Array.empty[Class[_]], Array.empty[Object])
  }

  def create[S <: Sentence](tagger: Class[_], argTypes: Array[Class[_]], argValues: Array[Object]): Constraint[S] = {
    try {
      val constructor = tagger.getConstructor(argTypes: _*)
      constructor.newInstance(argValues).asInstanceOf[Constraint[S]]
    } catch {
      case e: Exception =>
        throw new IllegalArgumentException("Could not create class: " + tagger.getName(), e);
    }
  }

  def getConstraintClass(classname: String): Class[_] = {
    Class.forName(classOf[Constraint[_]].getPackage().getName() + "." + classname)
  }

  def create[S <: Sentence](classname: String, argTypes: Array[Class[_]], argValues: Array[Object]): Constraint[S] = {
    create(getConstraintClass(classname), argTypes, argValues);
  }
}
